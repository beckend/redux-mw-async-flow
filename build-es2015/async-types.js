"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.defaultTypes = {
    ABORTED: '_ABORTED',
    END: '_END',
    FULFILLED: '_FULFILLED',
    PENDING: '_PENDING',
    REJECTED: '_REJECTED',
    REQUEST: '_REQUEST',
};
exports.getAsyncTypeConstants = ({ types } = {}) => (Object.assign({}, exports.defaultTypes, types));
exports.generateAsyncAction = ({ actionName, types = {} }) => {
    const { ABORTED, END, FULFILLED, PENDING, REJECTED, REQUEST, } = exports.getAsyncTypeConstants({ types });
    const ABORTED_NAME = `${actionName}${ABORTED}`;
    const END_NAME = `${actionName}${END}`;
    const FULFILLED_NAME = `${actionName}${FULFILLED}`;
    const PENDING_NAME = `${actionName}${PENDING}`;
    const REJECTEDT_NAME = `${actionName}${REJECTED}`;
    const REQUEST_NAME = `${actionName}${REQUEST}`;
    return {
        [ABORTED_NAME]: ABORTED_NAME,
        [END_NAME]: END_NAME,
        [FULFILLED_NAME]: FULFILLED_NAME,
        [PENDING_NAME]: PENDING_NAME,
        [REJECTEDT_NAME]: REJECTEDT_NAME,
        [REQUEST_NAME]: REQUEST_NAME,
    };
};
exports.replaceSuffix = (targetStr, suffixToReplace, replacementString) => `${targetStr.substring(0, targetStr.length - suffixToReplace.length)}${replacementString}`;
//# sourceMappingURL=async-types.js.map