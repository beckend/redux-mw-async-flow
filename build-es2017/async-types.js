"use strict";
const tslib_1 = require("tslib");
exports.defaultTypes = {
    REQUEST: 'REQUEST',
    PENDING: 'PENDING',
    FULFILLED: 'FULFILLED',
    REJECTED: 'REJECTED',
    ABORTED: 'ABORTED',
};
exports.getAsyncTypeConstants = ({ types } = {}) => {
    const { REQUEST, PENDING, FULFILLED, REJECTED, ABORTED, } = tslib_1.__assign({}, exports.defaultTypes, types);
    return {
        _REQUEST: `_${REQUEST}`,
        _PENDING: `_${PENDING}`,
        _FULFILLED: `_${FULFILLED}`,
        _REJECTED: `_${REJECTED}`,
        _ABORTED: `_${ABORTED}`,
    };
};
//# sourceMappingURL=async-types.js.map