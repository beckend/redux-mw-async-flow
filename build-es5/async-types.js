"use strict";
var tslib_1 = require("tslib");
exports.defaultTypes = {
    REQUEST: 'REQUEST',
    PENDING: 'PENDING',
    FULFILLED: 'FULFILLED',
    REJECTED: 'REJECTED',
    ABORTED: 'ABORTED',
};
exports.getAsyncTypeConstants = function (_a) {
    var types = (_a === void 0 ? {} : _a).types;
    var _b = tslib_1.__assign({}, exports.defaultTypes, types), REQUEST = _b.REQUEST, PENDING = _b.PENDING, FULFILLED = _b.FULFILLED, REJECTED = _b.REJECTED, ABORTED = _b.ABORTED;
    return {
        _REQUEST: "_" + REQUEST,
        _PENDING: "_" + PENDING,
        _FULFILLED: "_" + FULFILLED,
        _REJECTED: "_" + REJECTED,
        _ABORTED: "_" + ABORTED,
    };
};
//# sourceMappingURL=async-types.js.map