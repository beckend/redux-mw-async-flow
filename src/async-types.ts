export interface IDefaultTypes {
  readonly REQUEST: string;
  readonly PENDING: string;
  readonly FULFILLED: string;
  readonly REJECTED: string;
  readonly ABORTED: string;
  readonly END: string;
}
export const defaultTypes: IDefaultTypes = {
  REQUEST: '_REQUEST',
  PENDING: '_PENDING',
  FULFILLED: '_FULFILLED',
  REJECTED: '_REJECTED',
  ABORTED: '_ABORTED',
  END: '_END',
};

export type TDefaultTypesOptional = {
  readonly[P in keyof IDefaultTypes]?: IDefaultTypes[P];
};
export interface IGetAsyncTypeConstantsParams {
  readonly types?: TDefaultTypesOptional;
}
export const getAsyncTypeConstants = ({ types }: IGetAsyncTypeConstantsParams = {}) => (
  {
    ...defaultTypes,
    ...types,
  }
);

export const replaceSuffix = (targetStr: string, suffixToReplace: string, replacementString: string) =>
  `${targetStr.substring(0, targetStr.length - suffixToReplace.length)}${replacementString}`;

export interface IGetMetaResultTypes {
  readonly PENDING: string;
  readonly FULFILLED: string;
  readonly REJECTED: string;
  readonly ABORTED: string;
  readonly END: string;
}
export interface IGetMetaResultOverrides {
  readonly PAYLOAD: any;
  readonly ERROR: boolean;
  readonly[keyName: string]: boolean | string;
}
export const getMetaResult = (types: IGetMetaResultTypes, overrides: IGetMetaResultOverrides) => (
  {
    [types.PENDING]: false,
    [types.FULFILLED]: false,
    [types.REJECTED]: false,
    [types.END]: false,
    PAYLOAD: null,
    ERROR: null,
    ...overrides,
  }
);
