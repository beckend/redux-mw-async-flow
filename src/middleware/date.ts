/**
 * Separate module so it can be mocked
 */
export const newDate = () => new Date();
