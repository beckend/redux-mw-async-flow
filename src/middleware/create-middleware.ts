/* tslint:disable: no-unnecessary-local-variable */
/* tslint:disable: max-func-body-length */
/**
 * Detects action types with names ending with REQUEST, REJECTED, FULFILLED, ABORTED
 * Will dispatch PENDING if REQUEST
 * Will resolve if FULFILLED
 * REJECT when REJECTED or ABORTED
 * END after FULFILLED/REJECTED/ABORTED
 */
import { Middleware, Dispatch } from 'redux';
import { Action } from 'redux-actions';
import * as Bluebird from 'bluebird';
import * as lodash from 'lodash';
import {
  RequestStore,
  IRequestMap,
  REQUEST_KEY_PROMISE,
  REQUEST_KEY_RESOLVEFN,
  REQUEST_KEY_REJECTFN,
} from '../request-store';
import {
  getAsyncTypeConstants,
  replaceSuffix,
  TDefaultTypesOptional,
} from '../async-types';
import { createPromise } from '../promise-factory';
import { createObservers } from './middleware-observers';

const merge: typeof lodash.merge = require('lodash.merge');
const lSet: typeof lodash.set = require('lodash.set');
const lGet: typeof lodash.get = require('lodash.get');
const uniqueid = require('uniqueid');

const asyncUniqueId = uniqueid(null, '-@@ASYNC_FLOW');

// Base and full options
export interface IAsyncFlowActionMetaBase<TPayload> {
  // enable middleware, disabled will only acts as passthrough
  readonly enable: boolean;
  // the global timeout
  readonly timeout: number;
  // override timeout only for this action
  readonly timeoutRequest?: number;
  readonly promise: Bluebird<TPayload>;
}
// Added by middleware
export interface IAsyncFlowActionMetaAdded<TPayload> {
  readonly enable?: boolean;
  readonly timeout: number;
  readonly timeoutRequest: number;
  readonly promise: Bluebird<TPayload>;
}
// Optional version to pass what you want
export type TAsyncFlowActionMetaOptional<TActionPayload> = {
  readonly[P in keyof IAsyncFlowActionMetaBase<TActionPayload>]?: IAsyncFlowActionMetaBase<TActionPayload>[P];
}
// Can only apply to default metaKey
export interface IAsyncFlowActionOptional<TActionPayload> extends Action<TActionPayload> {
  meta: {
    asyncFlow: TAsyncFlowActionMetaOptional<TActionPayload>;
  };
}
// Action dispatched with meta added by middleware
export interface IAsyncFlowAction<TActionPayload> extends Action<TActionPayload> {
  meta: {
    asyncFlow: IAsyncFlowActionMetaAdded<TActionPayload>;
  };
}
export interface IGenerateIdFn<TAction> {
  (opts: { action: TAction }): string;
}
export interface IDefaultOpts<TAction> {
  // main key under meta
  readonly metaKey: string;
  // id key under main key
  readonly metaKeyRequestID: string;
  // timeout for promise before rejecting
  readonly timeout: number;
  // function to override generation of async ids
  readonly generateId: IGenerateIdFn<TAction>;
}
export const defaultOpts: IDefaultOpts<any> = {
  metaKey: 'asyncFlow',
  metaKeyRequestID: 'REQUEST_ID',
  timeout: 10000,
  generateId: ({ action }: { action: Action<any> }) => `${asyncUniqueId()}--${action.type}`,
};
export type TRequestId = string;
export type TDefaultOptsOptional<TAction> = {
  readonly[P in keyof IDefaultOpts<TAction>]?: IDefaultOpts<TAction>[P];
}
export interface ICreateAsyncFlowMiddlewareBaseOpts {
  // async constants
  asyncTypes?: TDefaultTypesOptional;
}
export type TCreateAsyncFlowMiddlewareOpts<TAction> = ICreateAsyncFlowMiddlewareBaseOpts & TDefaultOptsOptional<TAction>;
export const createAsyncFlowMiddleware = <TStoreState, TAction extends Action<any>>(opts: TCreateAsyncFlowMiddlewareOpts<TAction> = {
  metaKey: defaultOpts.metaKey,
  timeout: defaultOpts.timeout,
}) => {
  const {
    REQUEST,
    PENDING,
    FULFILLED,
    REJECTED,
    ABORTED,
    END,
  } = getAsyncTypeConstants({ types: opts.asyncTypes });

  const mwObservers = createObservers({
    asyncTypes: {
      REQUEST,
      END,
    },
  });

  const {
    metaKey,
    timeout,
    metaKeyRequestID,
    generateId,
  }: IDefaultOpts<TAction> = {
      ...defaultOpts,
      ...opts,
    };

  const requestStore = new RequestStore();

  const middleware: Middleware = () => {
    return (next: Dispatch<TStoreState>) => {
      return (action: TAction) => {
        /**
         * Normal dispatch without async
         */
        const dispatchNormal = () => next(action);
        const dispatchAsyncFlow = (actionArg: IAsyncFlowAction<any>) => {
          // Lets observers have a go before and after dispatches
          mwObservers.before.rootSubject.next(actionArg);
          const dispatchResult = next(actionArg);
          mwObservers.after.rootSubject.next(dispatchResult);
        };
        const actionType = action.type;
        /**
         * If meta.asyncFlow.enable is explicit set to false, completely skip this middleware.
         * The rest is up to the user to dispatch
         * REJECTED and FULFILLED suffixes manually
         */
        if (lGet<boolean>(action, ['meta', metaKey, 'enable']) === false) {
          dispatchNormal();
          return;
        }

        // used to get deep in action object
        const metaRequestIdPath = ['meta', metaKey, metaKeyRequestID];
        // iF set will dispatch en END action
        let actionEnd;
        // Resole, reject and then set dispatch END payload
        const handleEndAction = (suffixType: string, resolve: boolean, payloadArg?: any) => {
          const requestID = lGet<TRequestId>(action, metaRequestIdPath);
          if (requestID) {
            if (resolve) {
              requestStore.resolve(requestID, payloadArg || action.payload);
            } else {
              requestStore.reject(requestID, payloadArg || action.payload);
            }
            actionEnd = merge({}, action, {
              type: replaceSuffix(actionType, suffixType, END),
            });
          } else {
            console.warn(`${action.type} - meta data not found, did you forget to send it?`);
          }
        };
        /**
         * handle REQUEST, dispatches PENDING then dispatches REQUEST with meta data
         */
        if (actionType.endsWith(REQUEST)) {
          /**
           * Generate uniqueid, make sure it does not exist
           */
          let requestID = lGet<TRequestId>(action, metaRequestIdPath);
          if (!requestID || {}.hasOwnProperty.call(requestStore, requestID)) {
            do {
              requestID = generateId({ action });
            } while ({}.hasOwnProperty.call(requestStore, requestID));

            lSet<TRequestId>(action, metaRequestIdPath, requestID);
          }

          // User override of the timeout for each request
          const metaTimeoutKey = lGet<number>(action, ['meta', metaKey, 'timeoutRequest']);
          const timeoutRequest = metaTimeoutKey || timeout;
          /**
           * Dispatch pending first, with a twist, send a promise resolve reject with it
           * So this way user can use to the promise to do stuff after the action
           * The promise will be resolved later when getting ABORTED, REJECTED, or FULFILLED
           */
          const {
            promise,
            reject,
            resolve,
          } = createPromise<any>();
          promise
            .timeout(timeoutRequest, 'timeout')
            // .catch((er) => {
            //   // Dispatch REJECTED when TimeoutError
            //   if (er instanceof Bluebird.TimeoutError) {
            //     handleEndAction(REJECTED, false, er);
            //   } else {
            //     throw er || new Error('Unknown error');
            //   }
            // })
            .finally(() => {
              // Cleanup from requestStore
              requestStore.delete(requestID);
            });

          const tmpRequestStoreAddPayload: IRequestMap<any> = {
            [REQUEST_KEY_PROMISE]: promise,
            [REQUEST_KEY_RESOLVEFN]: resolve,
            [REQUEST_KEY_REJECTFN]: reject,
          } as any;
          // Register promise to requestStore
          requestStore.add(requestID, tmpRequestStoreAddPayload);
          /**
           * dispatch PENDING with meta
           */
          const addedActionMetaData = {
            meta: {
              [metaKey]: {
                timeout,
                timeoutRequest,
                promise,
              } as IAsyncFlowActionMetaAdded<any>
            }
          };
          const pendingAction: IAsyncFlowAction<any> = merge(
            {},
            action,
            {
              type: replaceSuffix(actionType, REQUEST, PENDING),
            },
            addedActionMetaData
          ) as any;
          dispatchAsyncFlow(pendingAction);

          /**
           * Dispatch orginal action with meta data
           */
          const newAction: IAsyncFlowAction<any> = merge({}, action, addedActionMetaData) as any;
          dispatchAsyncFlow(newAction);
          return;
        } else if (actionType.endsWith(FULFILLED)) {
          handleEndAction(FULFILLED, true);
        } else if (actionType.endsWith(REJECTED)) {
          handleEndAction(REJECTED, false);
        } else if (actionType.endsWith(ABORTED)) {
          handleEndAction(ABORTED, false);
        } else {
          /**
           * Normal dispatch when unmatched by suffixes
           */
          dispatchNormal();
          return;
        }

        // Reaching here it's atually a IAsyncFlowAction
        dispatchAsyncFlow(action as any);
        /**
         * Dispatch actionEnd if set for completion
         */
        if (actionEnd) { dispatchAsyncFlow(actionEnd); }
      };
    };
  };

  return {
    middleware,
    observers: mwObservers,
  };
};
