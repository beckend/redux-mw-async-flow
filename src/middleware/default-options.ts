import { IDefaultOpts } from './create-middleware';

export const defaultOpts: IDefaultOpts<any> = {
  metaKey: 'asyncFlow',
  metaKeyRequestID: 'REQUEST_ID',
  timeout: 10000,
};
