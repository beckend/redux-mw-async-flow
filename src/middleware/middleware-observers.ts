/**
 * Rxjs observers only triggered when the async action is matched/enabled
 */
import { Observable } from 'rxjs/Observable';
import { filter } from 'rxjs/operator/filter';
import { Subject } from 'rxjs/Subject';
import '../rxjs/add/__invoke';
import { IAsyncFlowAction } from './create-middleware';
export { Observable };

export interface IAsyncTypes {
  readonly REQUEST: string;
  readonly END: string;
}
export interface ICreateObserversArgs {
  // async constants
  readonly asyncTypes: IAsyncTypes;
}
export const createObservers = ({ asyncTypes }: ICreateObserversArgs) => {
  const createAllObservers = () => {
    // Gets all the actions through the middleware
    const rootSubject = new Subject<IAsyncFlowAction<any>>();

    // lock out .next api etc
    const obsOnAll = rootSubject.asObservable();

    // filter by requests
    const obsOnRequest: Observable<IAsyncFlowAction<any>> = obsOnAll
      .__invoke(
        filter, (action: IAsyncFlowAction<any>) => action.type.endsWith(asyncTypes.REQUEST)
      );

    const obsOnEnd: Observable<IAsyncFlowAction<any>> = obsOnAll
      .__invoke(
        filter, (action: IAsyncFlowAction<any>) => action.type.endsWith(asyncTypes.END)
      );

    return {
      obsOnAll,
      obsOnEnd,
      obsOnRequest,
      rootSubject,
    };
  };

  return {
    // after dispatch
    after: createAllObservers(),
    // before dispatch
    before: createAllObservers(),
  };
};
