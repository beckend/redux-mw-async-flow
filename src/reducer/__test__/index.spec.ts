/* tslint:disable: chai-vague-errors */
import {
  createAsyncFlowReducer,
  IState
} from '..';
import { ReducerMeta } from 'redux-actions';
import { defaultOpts as middlewareDefaultOpts } from '../../middleware/default-options';
import { defaultTypes } from '../../async-types';

describe('reducer creation', () => {
  it('can create without options', () => {
    const reducer = createAsyncFlowReducer();
    const {
      metaKey,
      metaKeyRequestID
    } = middlewareDefaultOpts;
    expect(reducer(undefined as any, {
      type: 'MY_ACTION_REQUEST',
      [metaKey]: {
        [metaKeyRequestID]: 'myId'
      }
    } as any))
      .toMatchSnapshot();
  });

  it('can create with options and works', () => {
    const asyncRequestSuffix = '-request-suffix';
    const metaKey = '__THE_META_KEY';
    const metaKeyRequestID = '__THE_META_REQ_ID';
    const reducer = createAsyncFlowReducer({
      asyncTypes: {
        REQUEST: asyncRequestSuffix,
        PENDING: '-pending-SUFF',
        FULFILLED: 'fulfilledsuFf',
        REJECTED: '____rejected___-',
        ABORTED: '--__aborted-ab'
      },
      metaKey,
      metaKeyRequestID
    });

    expect(reducer(undefined as any, {
      type: `something${asyncRequestSuffix}`,
      [metaKey]: {
        [metaKeyRequestID]: 'myId'
      }
    } as any))
      .toMatchSnapshot();
  });
});

describe('reducer tests', () => {
  let reducer: ReducerMeta<IState, any, any>;
  const {
    metaKey,
    metaKeyRequestID
  } = middlewareDefaultOpts;
  beforeEach(() => {
    reducer = createAsyncFlowReducer();
  });

  it('initial state, with any action works', () => {
    expect(reducer(undefined as any, {
      type: 'MYACTION'
    } as any))
      .toMatchSnapshot();
  });

  Object.values(defaultTypes).forEach((asyncSuffix) => {
      it(`handle ${asyncSuffix}`, () => {
        const firstState = reducer(undefined as any, {
          type: `MYACTION${asyncSuffix}`,
          [metaKey]: {
            [metaKeyRequestID]: 'hello'
          }
        } as any);
        expect(firstState)
          .toMatchSnapshot();
        expect(reducer(firstState, {
          type: `MYACTION${asyncSuffix}`,
          [metaKey]: {
            [metaKeyRequestID]: 'hello2'
          }
        } as any))
          .toMatchSnapshot();
      });
  });
});
