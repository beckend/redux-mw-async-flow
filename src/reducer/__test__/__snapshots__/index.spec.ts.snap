// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducer creation can create with options and works 1`] = `
Object {
  "counters": Object {
    "--__aborted-ab": 0,
    "-pending-SUFF": 0,
    "-request-suffix": 1,
    "_END": 0,
    "____rejected___-": 0,
    "fulfilledsuFf": 0,
  },
  "latestActions": Object {
    "something-request-suffix": Object {
      "__THE_META_KEY": Object {
        "__THE_META_REQ_ID": "myId",
      },
      "type": "something-request-suffix",
    },
  },
}
`;

exports[`reducer creation can create without options 1`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 1,
  },
  "latestActions": Object {
    "MY_ACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "myId",
      },
      "type": "MY_ACTION_REQUEST",
    },
  },
}
`;

exports[`reducer tests handle _ABORTED 1`] = `
Object {
  "counters": Object {
    "_ABORTED": 1,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello",
      },
      "type": "MYACTION_ABORTED",
    },
  },
}
`;

exports[`reducer tests handle _ABORTED 2`] = `
Object {
  "counters": Object {
    "_ABORTED": 2,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello2",
      },
      "type": "MYACTION_ABORTED",
    },
  },
}
`;

exports[`reducer tests handle _END 1`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 1,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello",
      },
      "type": "MYACTION_END",
    },
  },
}
`;

exports[`reducer tests handle _END 2`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 2,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello2",
      },
      "type": "MYACTION_END",
    },
  },
}
`;

exports[`reducer tests handle _FULFILLED 1`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 1,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello",
      },
      "type": "MYACTION_FULFILLED",
    },
  },
}
`;

exports[`reducer tests handle _FULFILLED 2`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 2,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello2",
      },
      "type": "MYACTION_FULFILLED",
    },
  },
}
`;

exports[`reducer tests handle _PENDING 1`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 1,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello",
      },
      "type": "MYACTION_PENDING",
    },
  },
}
`;

exports[`reducer tests handle _PENDING 2`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 2,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello2",
      },
      "type": "MYACTION_PENDING",
    },
  },
}
`;

exports[`reducer tests handle _REJECTED 1`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 1,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello",
      },
      "type": "MYACTION_REJECTED",
    },
  },
}
`;

exports[`reducer tests handle _REJECTED 2`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 2,
    "_REQUEST": 0,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello2",
      },
      "type": "MYACTION_REJECTED",
    },
  },
}
`;

exports[`reducer tests handle _REQUEST 1`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 1,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello",
      },
      "type": "MYACTION_REQUEST",
    },
  },
}
`;

exports[`reducer tests handle _REQUEST 2`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 2,
  },
  "latestActions": Object {
    "MYACTION_REQUEST": Object {
      "asyncFlow": Object {
        "REQUEST_ID": "hello2",
      },
      "type": "MYACTION_REQUEST",
    },
  },
}
`;

exports[`reducer tests initial state, with any action works 1`] = `
Object {
  "counters": Object {
    "_ABORTED": 0,
    "_END": 0,
    "_FULFILLED": 0,
    "_PENDING": 0,
    "_REJECTED": 0,
    "_REQUEST": 0,
  },
  "latestActions": Object {},
}
`;
